docs_src/design/adr/device-service/0021-invoking-functions.md

```diff
@@ -0,0 +1,112 @@
+# Invoking Device Functions
+ 
+
+## Status
+
+**Draft**
+
+## Context
+This ADR presents a mechanism for invoking functions on a device.
+
+## Existing Behavior
+
+Device access in EdgeX is focussed on 'Readings' and 'Settings'. A device may
+present values which are readable, representing some measurement or status,
+such as a temperature or vibration reading. Also a setting such as a the
+required speed of a fan motor may be written to and read back.
+
+What is not supported directly is 'Functions', where a device may be commanded
+to do something. Examples could include perform a self-test, go into standby
+mode for an hour, invalidate an access key.
+
+While such operations may modelled using virtual resources in a device profile,
+this is unintuitive.
+
+## Decision
+
+**Add a new section to device profiles describing functions**
+
+```
+{
+  "deviceFunctions":
+  [
+    {
+      "name": "Name by which the function is accessed",
+      "description": "Readable description of the function",
+      "attributes": { device-service-specific attributes which select this function },
+      "parameters":
+      [
+        {
+          "name": "Parameter name",
+          "description": "(optional) description of what the parameter controls",
+          "type": "Any of the usual EdgeX data types",
+          "defaultValue": "(optional) value to use if param is not supplied",
+          "maximum": "(optional) for numerics, maximum allowed value",
+          "minimum": "(optional) for numerics, minimum allowed value"
+        }
+      ],
+      "returnValues":
+      [
+        {
+          "name": "Name of returned value",
+          "description": "(optional) description of what the value indicates",
+          "type": "Any of the usual EdgeX data types"
+        }
+      ]
+    }
+  ]
+}
+```
+
+Note: the `attributes` structure is analagous to `attributes` in a `deviceResource`. Each device service should document and implement a scheme of required attributes that will allow for selection of the relevant funtion.
```

Comment:
I'm not sure what the purpose of attributes are in this context? If extra data is required to make a function call, isn't this the purpose of parameters? I also don't understand what you mean by "a scheme of required attributes that will allow for selection of relevant function"?

sp: funtion

I